### PR TITLE
Parse annotations one time

### PR DESCRIPTION
**Pull request by nstrelow**
* Parse annotations once for each file and construct a tuple: image_path, bbox, label
* The annotations array of tuples is then split into train and test splits and stored inside the ImageBoundingBoxDataset.

Appropriate renaming was done, but I am open to better naming suggestions

### CODE REVIEW

File: multibuildingdetector/readers/XMLReader.py [hunk 1]
```diff
@@ -1,8 +1,13 @@
...
+from os.path import isfile, join
 import xml.etree.ElementTree as ET
-import chainer
+from sklearn.model_selection import train_test_split
 from chainercv.utils import read_image
 
+from .ImageBoundingBoxDataset import ImageBoundingBoxDataset
+
+
 LABEL_NAMES = ('other',
```
Comment (fapaul) [inline multibuildingdetector/readers/XMLReader.py]: Since we have now separated dataset and reader, I think the label_names belong to the dataset, hence we can put this in the dataset file.

**Review by fapaul — CHANGES_REQUESTED** [review]
Do we have a performance gain through these changes?

File: multibuildingdetector/readers/XMLReader.py (hunk 1 above)
Comment (nstrelow) [reply multibuildingdetector/readers/XMLReader.py]: But the labels are specific to what we have and how we read it, hence the it's in the reader.

Therefore other implementation can use the same dataset, but a different reader.
Actually this belongs in a config file.

Comment (fapaul) [reply multibuildingdetector/readers/XMLReader.py]: I see but we could just return `train, test` in the reader class and instantiate the `ImageBoundingBoxDataset` class in `__main__`. This way the users can add new reader classes as long as they implement something for the Dataset and you can add it in the yml config.

**Comment by nstrelow** [comment]
I will try to provide the amount of iterations we gain trough this.
The iters/sec vary very much on the server. Tried a bit switching between branches, but I could not get an accurate result. Could be due to the matplotlib overhead.

I will let it run for a while and report back.

**Comment by nstrelow** [comment]
But in theory performance should be better
```
    def get_example(self, i):
        img_file, bbox, label = self.annotations[i]
        img = read_image(img_file, color=True)
        return img, bbox, label
```
The get example gets executed each time an image with annotation is used to train.
Here we only read from an array, unpack a tuple, read an image and return a tuple.

In earlier implementations the annotation .xml file was loaded and parsed with ET.

**Comment by nstrelow** [comment]
It is slightly better, but not much. 0.05 iters/sec. I think this is still due to other things slowing us down like matplotlib.

I would add this anyway, as it should save us runtime anyway.

**Comment by fapaul** [comment]
Since we decided to skip the colour conversion for now. Can you make the performance tests without it again, please?

**Comment by fapaul** [comment]
@nstrelow Please review new structure

**Comment by nstrelow** [comment]
We would need to remove the color conversion or replace it with opencv again

**Comment by nstrelow** [comment]
@mockingbird2 The new structure is nice, but right now the parser is very specific.
It needs to return a tuple with 3 values which is compatible with ImageBoundingBoxDataset.

But if we say that every parser needs to return compatible types, then there should be no problem.

**Comment by fapaul** [comment]
@nstrelow I'd suggest removing the colour conversion for now. Especially when the opencv version slows us down drastically.

**Comment by nstrelow** [comment]
@mockingbird2 Ok, I will remove it from here.
I do not believe opencv slows us that down as we initially thought. No matter which color conversion we use our iters/sec still drop first. I will test it with and without opencv and report back.

**Review by fapaul — APPROVED** [review]
